From ac392a9a206338129eac1392e5d247042f3cb3fd Mon Sep 17 00:00:00 2001
From: Caroline <cljin@mit.edu>
Date: Sun, 30 Oct 2022 20:57:28 -0400
Subject: [PATCH] added milliseconds to filename

---
 doc/muxers.texi     |  3 +++
 libavformat/utils.c | 26 ++++++++++++++++++++++++--
 2 files changed, 27 insertions(+), 2 deletions(-)

diff --git a/doc/muxers.texi b/doc/muxers.texi
index 4edbb22b00..2400859f2e 100644
--- a/doc/muxers.texi
+++ b/doc/muxers.texi
@@ -1392,6 +1392,9 @@ If the pattern contains "%d" or "%0@var{N}d", the first filename of
 the file list specified will contain the number 1, all the following
 numbers will be sequential.
 
+The pattern can also contain "%t" which writes out the timestamp of the
+1388	frame in the format "HH.mm.ss.fff"
+
 The pattern may contain a suffix which is used to automatically
 determine the format of the image files to write.
 
diff --git a/libavformat/utils.c b/libavformat/utils.c
index cf4d68bff9..dbe76d0e30 100644
--- a/libavformat/utils.c
+++ b/libavformat/utils.c
@@ -295,11 +295,15 @@ int av_get_frame_filename2(char *buf, int buf_size, const char *path, int number
 {
     const char *p;
     char *q, buf1[20], c;
-    int nd, len, percentd_found;
+    int nd, len, percentd_found, percentt_found;
+    int64_t unow, ms;
+    time_t now;
+    struct tm tm;
 
     q = buf;
     p = path;
     percentd_found = 0;
+    percentt_found = 0;
     for (;;) {
         c = *p++;
         if (c == '\0')
@@ -331,6 +335,24 @@ int av_get_frame_filename2(char *buf, int buf_size, const char *path, int number
                 memcpy(q, buf1, len);
                 q += len;
                 break;
+            case 't':
+                if (percentd_found)
+                    goto fail;
+                // copy from vf_draw to get the time to precision of ms
+                // func_strftime in vf_drawtext.c
+                percentt_found = 1;
+                unow = av_gettime();
+                ms = (unow/1000)%1000;
+                now  = unow / 1000000;
+                localtime_r(&now, &tm);
+                snprintf(buf1, sizeof(buf1),
+                       "%02d.%02d.%02d.%03lld", tm.tm_hour, tm.tm_min, tm.tm_sec, ms);
+                len = strlen(buf1);
+                if ((q - buf + len) > buf_size - 1)
+                    goto fail;
+                memcpy(q, buf1, len);
+                q += len;
+                break;
             default:
                 goto fail;
             }
@@ -340,7 +362,7 @@ addchar:
                 *q++ = c;
         }
     }
-    if (!percentd_found)
+    if (!(percentd_found || percentt_found))
         goto fail;
     *q = '\0';
     return 0;
-- 
2.32.1 (Apple Git-133)

